pkg/reconciler/revision/background.go

```diff
@@ -157,53 +157,71 @@ func (r *backgroundResolver) Resolve(logger *zap.SugaredLogger, rev *v1.Revision
 		r.addWorkItems(rev, name, opt, registriesToSkip, timeout)
-		return nil, nil
+		return nil, nil, nil
 	}
 
 	if !result.ready() {
 		logger.Debug("Resolve request in flight, returning nil, nil")
-		return nil, nil
+		return nil, nil, nil
 	}
 
 	ret := r.results[name]
 	if ret.err != nil {
 		logger.Debugf("Resolve returned the resolved error: %v", ret.err)
-		return nil, ret.err
+		return nil, nil, ret.err
 	}
 
-	array := make([]v1.ContainerStatus, len(rev.Spec.Containers))
-	for k, v := range ret.statuses {
-		array[k] = v
+	initContainerStatuses = make([]v1.ContainerStatus, len(rev.Spec.InitContainers))
+	for i, container := range rev.Spec.InitContainers {
+		if resolvedDigest, ok := ret.imagesResolved[container.Image]; ok {
+			initContainerStatuses[i] = v1.ContainerStatus{
+				Name:        container.Name,
+				ImageDigest: resolvedDigest,
+			}
+		}
+	}
+
+	statuses = make([]v1.ContainerStatus, len(rev.Spec.Containers))
+	for i, container := range rev.Spec.Containers {
+		if resolvedDigest, ok := ret.imagesResolved[container.Image]; ok {
+			statuses[i] = v1.ContainerStatus{
+				Name:        container.Name,
+				ImageDigest: resolvedDigest,
+			}
+		}
 	}
 
-	logger.Debugf("Resolve returned %d resolved images for revision", len(array))
-	return array, nil
+	logger.Debugf("Resolve returned %d resolved images for revision", len(statuses)+len(initContainerStatuses))
+	return initContainerStatuses, statuses, nil
 }
 
 // addWorkItems adds a digest resolve item to the queue for each container in the revision.
 // This is expected to be called with the mutex locked.
 func (r *backgroundResolver) addWorkItems(rev *v1.Revision, name types.NamespacedName, opt k8schain.Options, registriesToSkip sets.String, timeout time.Duration) {
+	totalNumOfContainers := len(rev.Spec.Containers) + len(rev.Spec.InitContainers)
 	r.results[name] = &resolveResult{
-		opt:              opt,
-		registriesToSkip: registriesToSkip,
-		statuses:         make(map[int]v1.ContainerStatus),
-		want:             len(rev.Spec.Containers),
-		workItems:        make([]workItem, len(rev.Spec.Containers)),
+		opt:                opt,
+		registriesToSkip:   registriesToSkip,
+		imagesResolved:     make(map[string]string),
+		imagesToBeResolved: sets.String{},
+		workItems:          make([]workItem, 0, totalNumOfContainers),
 		completionCallback: func() {
 			r.enqueue(name)
 		},
 	}
 
-	for i, container := range rev.Spec.Containers {
+	for _, container := range append(rev.Spec.InitContainers, rev.Spec.Containers...) {
+		r.results[name].imagesToBeResolved.Insert(container.Image)
+	}
+
+	for _, container := range append(rev.Spec.InitContainers, rev.Spec.Containers...) {
 		item := workItem{
 			revision: name,
 			timeout:  timeout,
 			name:     container.Name,
 			image:    container.Image,
-			index:    i,
 		}
-
-		r.results[name].workItems[i] = item
+		r.results[name].workItems = append(r.results[name].workItems, item)
 		r.queue.AddRateLimited(item)
```

Comment:
if two containers use the same image, will we end up with two workItems both resolving that image, and if so is that intended/safe? (e.g. you could have one complete after the overall revision has already reported success/error). Could we only add one work item for each image maybe?

Reply:
That was the intention to keep the same semantics as before when adding items to the queue. Didnt want to break that.
Might be better to optimize it. Will test.